include/beast/core/detail/clamp.hpp

```diff
@@ -0,0 +1,40 @@
+//
+// Copyright (c) 2013-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_CORE_DETAIL_CLAMP_HPP
+#define BEAST_CORE_DETAIL_CLAMP_HPP
+
+#include <limits>
+#include <cstdlib>
+
+namespace beast {
+namespace detail {
+
+template<class UInt>
+static
+std::size_t
+clamp(UInt x)
+{
+    if(x >= std::numeric_limits<std::size_t>::max())
```

Comment:
There are places in the WebSocket protocol which specify `uint64_t`. For example, the size of a frame can be 2^64-1 bytes. However, internally we use `std::size_t` to represent sizes which we might need to allocate. Clearly, an executable compiled for a 32-bit address model cannot allocate more than 2^32-1. This function exists to clamp a value that will be used for some type of computation involving a buffer. On 32 bit platforms it converts `std::uint64_t` to `std::uint32_t` by clamping. On 64 bit platforms it ends up doing nothing.